### PR TITLE
Avoid rebuilding of parser targets

### LINKED ISSUE #413: make / make install re-compiles some parser files without code changes
`make install` shouldn't re-compile code but what I see is:

```
$ make install
[  1%] Built target pyembed
[  3%] Built target fmt
[  4%] Built target test_util
[  5%] Built target pyastgen
[  6%] -- NMODL : GENERATING DIFFERENTIAL EQUATION PARSER WITH BISON! --
/gpfs/bbp.cscs.ch/data/project/proj16/kumbhar/pramod_scratch/nmodl/src/parser/diffeq.yy: warning: 5 shift/reduce conflicts [-Wconflicts-sr]
[  6%] Building CXX object src/lexer/CMakeFiles/lexer_obj.dir/diffeq_base_lexer.cpp.o
[  7%] -- NMODL : GENERATING UNIT PARSER WITH BISON! --
/gpfs/bbp.cscs.ch/data/project/proj16/kumbhar/pramod_scratch/nmodl/src/parser/unit.yy: warning: 1 shift/reduce conflict [-Wconflicts-sr]
[  8%] Building CXX object src/lexer/CMakeFiles/lexer_obj.dir/unit_base_lexer.cpp.o
[  9%] Building CXX object src/lexer/CMakeFiles/lexer_obj.dir/__/parser/diffeq_driver.cpp.o
[  9%] Building CXX object src/lexer/CMakeFiles/lexer_obj.dir/__/parser/diffeq_context.cpp.o
[ 10%] Building CXX object src/lexer/CMakeFiles/lexer_obj.dir/__/parser/unit_driver.cpp.o
[ 10%] Building CXX object src/lexer/CMakeFiles/lexer_obj.dir/__/parser/diffeq/diffeq_parser.cpp.o
[ 10%] -- NMODL : GENERATING C (11) PARSER WITH BISON! --
[ 11%] Building CXX object src/lexer/CMakeFiles/lexer_obj.dir/__/parser/c/c11_parser.cpp.o
[ 12%] Building CXX object src/lexer/CMakeFiles/lexer_obj.dir/__/parser/unit/unit_parser.cpp.o
[ 23%] Built target lexer_obj
[ 24%] Linking CXX static library liblexer.a
..
```

and doing it again:

```
s.ch/data/project/proj16/kumbhar/pramod_scratch/nmodl/build$ make install
[  1%] Built target pyembed
[  3%] Built target fmt
[  4%] Built target test_util
[  5%] Built target pyastgen
[  6%] -- NMODL : GENERATING DIFFERENTIAL EQUATION PARSER WITH BISON! --
/gpfs/bbp.cscs.ch/data/project/proj16/kumbhar/pramod_scratch/nmodl/src/parser/diffeq.yy: warning: 5 shift/reduce conflicts [-Wconflicts-sr]
[  6%] Building CXX object src/lexer/CMakeFiles/lexer_obj.dir/diffeq_base_lexer.cpp.o
[  7%] -- NMODL : GENERATING UNIT PARSER WITH BISON! --
/gpfs/bbp.cscs.ch/data/project/proj16/kumbhar/pramod_scratch/nmodl/src/parser/unit.yy: warning: 1 shift/reduce conflict [-Wconflicts-sr]
[  8%] Building CXX object src/lexer/CMakeFiles/lexer_obj.dir/unit_base_lexer.cpp.o
[  9%] Building CXX object src/lexer/CMakeFiles/lexer_obj.dir/__/parser/diffeq_driver.cpp.o
[  9%] Building CXX object src/lexer/CMakeFiles/lexer_obj.dir/__/parser/diffeq_context.cpp.o
[ 10%] Building CXX object src/lexer/CMakeFiles/lexer_obj.dir/__/parser/unit_driver.cpp.o
[ 10%] Building CXX object src/lexer/CMakeFiles/lexer_obj.dir/__/parser/diffeq/diffeq_parser.cpp.o
[ 10%] -- NMODL : GENERATING C (11) PARSER WITH BISON! --
[ 11%] Building CXX object src/lexer/CMakeFiles/lexer_obj.dir/__/parser/c/c11_parser.cpp.o
[ 12%] Building CXX object src/lexer/CMakeFiles/lexer_obj.dir/__/parser/unit/unit_parser.cpp.o
[ 23%] Built target lexer_obj
[ 24%] Linking CXX static library liblexer.a
....
```

### PR DESCRIPTION
**Pull request by pramodk**
- In #385 we used common position and location classes
    from nmodl parser
  - But output of CMake targets were not updated resulting
    in regenerating parser files every time

fixes #413

### CODE REVIEW

**Comment by alkino** [comment]
We don't use location.hh and position.hh?

File: src/lexer/CMakeLists.txt [hunk 1]
```diff
@@ -93,8 +93,6 @@ add_custom_command(
 add_custom_command(
   OUTPUT ${PROJECT_BINARY_DIR}/src/parser/diffeq/diffeq_parser.cpp
          ${PROJECT_BINARY_DIR}/src/parser/diffeq/diffeq_parser.hpp
-         ${PROJECT_BINARY_DIR}/src/parser/diffeq/location.hh
-         ${PROJECT_BINARY_DIR}/src/parser/diffeq/position.hh
          ${PROJECT_BINARY_DIR}/src/parser/diffeq/stack.hh
   COMMAND ${BISON_EXECUTABLE} ARGS -d -o ${PROJECT_BINARY_DIR}/src/parser/diffeq/diffeq_parser.cpp
           ${NMODL_PROJECT_SOURCE_DIR}/src/parser/diffeq.yy
```
Comment (alkino) [inline src/lexer/CMakeLists.txt:98]: We can clean that by using https://cmake.org/cmake/help/v3.8/module/FindBISON.html

Comment (pramodk) [reply src/lexer/CMakeLists.txt:98]: Do you mean using `BISON_TARGET`?

Comment (alkino) [reply src/lexer/CMakeLists.txt:98]: Yes

Comment (alkino) [reply src/lexer/CMakeLists.txt:98]: notice that `FLEX_TARGET` exists too

**Comment by iomaganaris** [comment]
> We don't use location.hh and position.hh?

I had the same question as well. The logic behind the deletion of the header file is that they will be the same for every execution of bison and this is the reason why we don't have to add them in the `OUTPUT` of multiple `add_custom_command`s?

File: src/lexer/CMakeLists.txt (hunk 1 above)
Comment (pramodk) [reply src/lexer/CMakeLists.txt:98]: @alkino : I looked at BISON_TARGET and it has form:

> BISON_TARGET(MyParser parser.y parser.cpp DEFINES_FILE parser.h)

what I am missing is that possibility to set dependency with other header files. e.g. In our case `diffeq.yy` depends on `diffeq_context.hpp` and `diffeq_helper.hpp`. I don't see straightforward/compact way to add extra dependencies to `MyParser` target.

So, I think its fine to keep the current custom_command.

**Comment by pramodk** [comment]
> We don't use location.hh and position.hh?
>  The logic behind the deletion of the header file is that they will be the same for every execution of bison and this is the reason why we don't have to add them in the OUTPUT of multiple add_custom_commands?

No, we do use them but from different parser (i.e. nmodl). In #385 we removed `%locations`. This means, for some parsers we are NOT generating location & position headers anymore. And that's why we have to remove those non-existent output files. Otherwise, cmake thinks that output files are missing and it regenerate parsers every time we run make.

**Review by alkino — APPROVED** [review]

**Pull request merged**